domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -105,6 +111,29 @@ class ExplorationProgressController @Inject constructor(
     }
   }
 
+  /**
+   * Checks if progress made in the exploration is saved and the checkpoint database has not
+   * exceeded limit.
+   *
+   * If progress is not saved, this function throws [ProgressNotSavedException].
+   * If checkpoint database has exceeded the allocated size limit this function throws
+   * [CheckpointDatabaseOverflowException]
+   */
+  internal fun isCurrentCheckpointStateIsSavedDatabaseNotExceededLimit() {
+    explorationProgressLock.withLock {
+      // throw an appropriate exception if either the checkpoint database has exceeded the allocated
+      // size limit or the current exploration contains unsaved progress.
+      val currentCheckpointState = explorationProgress.checkpointState
+      if (currentCheckpointState == CheckpointState.CHECKPOINT_SAVED_DATABASE_EXCEEDED_LIMIT) {
+        throw CheckpointDatabaseOverflowException(
```

Comment:
We shouldn't use exceptions in this way. Exceptions should only be used in exceptional cases. Here, we expect these errors to potentially happen which means returning an enum status would probably be clearer & avoids the potential confusion that these exceptions are catching actual errors vs. legitimate states.

Also, conceptually it's a successful state if the checkpoint can't be saved for some detected reason. AsyncResult's error state should also be exception/reserved for unexpected issues that are detected.

Reply:
changed the approach to dataProvider approach so this is above changes are no longer needed.